CIB-Mango-Tree/mango-tango-cli
# Time of Day Interval Analysis I

### Description

This merge request implements the foundational code for approaching ingesting and processing datetime data. It takes a
CSV and processes it as a Polars dataframe. Upon user input of a TIME__INTERVAL__LENGTH, the Polars dataframe is grouped into intervals of time based on that input and aggregated into counts of records. These time interval groupings and their counts are then plotted into bar graphs represented both via Plotly Express (HTML) and Plotext (CLI).

The following packages were added to the requirements.txt file to support this code MR:
- ipykernel==6.29.5
	- to support using Jupyter notebooks
- plotly==5.24.1
	- to create publishable graphs on HTML, also just one of my go-to visualization packages! (https://plotly.com/python/)
- pandas==2.2.3
	- to export Polars dataframes to for the purpose of Plotly visualizations
- pyarrow==17.0.0
	- ran into some dependency issues while testing Pandas, Plotly, and Plotext
- nbformat==5.10.4
	- for further Jupyter notebook support, ran into some dependency issues while testing
- plotext==5.3.2
	- to plot graphs on the command line (https://github.com/piccolomo/plotext)
	
_TLDR:_ Input is a CSV with datetime data. Output is a bar graph showing record counts grouped by time intervals. This is a very simple test intended to just begin the foundation of temporal analysis.
- input schema strict requirements: a datetime column that Polars can recognize

### Changes Made

- temporal folder: added in analyzers folder to hold data processing and analysis Python code
	- temporal.ipynb: a Python notebook used to test out code functions that were later translated to the main.py file
	- main.py: an executable Python file that takes a CSV from the directory, ingests it as a Polars dataframe, and returns a bar
	graph of the count of records in the dataset grouped by a user-determined time interval.
		- functions: load_csv, process_datetime_feature_engineering, analyze_time_of_day, plot_time_of_day_to_terminal, plot_time_of_day_to_plotly

### Related Issues

N/A; no tickets made

### How to Review this MR

1. Update your venv environment to the updated requirements.txt
	- pip install -r requirements.txt
2. Read through the functions written in main.py and the inline documentation to understand what each function does.
3. Run main.py to see if output (CLI graph, HTML graph, CSV export) is as expected. Play around with the inputs written in the if __name__ == '__main__' execution code.
	- python main.py
	- Alternatively, play around with the .ipynb file to see if functions work as expected.

You should expect outputs that look like this:
![image](https://github.com/user-attachments/assets/0e198157-f19e-44dd-a3cb-9c4857155550)
![image](https://github.com/user-attachments/assets/443e683d-e561-4e65-b280-1133c7192530)

### Technical Debt

- Properly integrate this MR's code with the existing mango-tango CLI and remove the __main__ execution code
   - which will also allow for export to other file formats
   - I considered using argparse to make CSV__INPUT and TIME__INTERVAL__LENGTH CLI inputs but thought that would be redundant work since the interface code has all been written out
- Build CLI functionality to choose between multiple datetime columns
- Iron out edge cases for x-tick labeling in the plotext CLI graphing (e.g. the first x-tick does not show when TIME__INTERVAL__LENGTH = 60)

### Future Improvements

- Automatically detect what CSVs users are able to run this test on
- Add new kinds of tables for tabular export (to be scoped out)
- Test on other data outside of reddit_vm.csv (need to load it into the repository first or point to the Google Drive data via Google API)
- Remove all functions definitions from the .ipynb and have it just import from the .py file as a single source of truth, just keep it for testing.
- Integrate the HTML export into dashboard integration (that can later be hosted)
- Combine the foundations of this with other more complex temporal-related tests
